larq/optimizers_v2.py

```diff
@@ -1,10 +1,153 @@
+import warnings
+
 import tensorflow as tf
 import larq as lq
 
 from larq import utils
 from copy import deepcopy
 
 
+@utils.register_keras_custom_object
+class CaseOptimizer(tf.keras.optimizers.Optimizer):
+    """An optmizer wrapper that applies different optimizers to a subset of variables.
+
+    An optimizer is used to train a variable iff its accompanying predicate evaluates to
+    `True`.
+
+    For each variable, at most one optimizer's predicate may evaluate to `True`. If no
+    optimizer's predicate evaluates to `True` for a variable, it is trained with the
+    `default` optimizer. If a variable is claimed by no optimizers and `default==None`,
+    the variable is not trained.
+
+    # Arguments
+    pred_opt_pairs: One or more `(pred, tf.keras.optimizers.Optimzer)` pairs, where 
+        `pred`  takes one `tf.Variable` as argument and returns `True` if the optimizer 
+        should be used for that variable, e.g. `pred(var) == True`.
+    default: A `tf.keras.optimizers.Optimizer` to be applied to any variable not claimed
+        by any other optimizer. (Must be passed as keyword argument.)
+    """
+
+    def __init__(self, *pred_opt_pairs, default=None, name="optimizer_case"):
```

Comment:
Do you think renaming `default` to `default_optimizer` would be more descriptive?

Reply:
`default` matches the `tf.case`, but I'd be good with calling it `default_opt` (to also match `pred_opt_pairs`). What do you think?

Reply:
Thinking of this again, what do you think about `predicate_optimizer_pairs` and `default_optimizer`? I think it is a bit long, but would be more descriptive.

Reply:
Done in e4695d2.